File: app/src/main/java/nl/rsdt/japp/jotial/maps/management/transformation/async/AsyncTransduceTask.java
package nl.rsdt.japp.jotial.maps.management.transformation.async;

import android.os.AsyncTask;

import nl.rsdt.japp.jotial.maps.management.transformation.AbstractTransducer;

/**
 * @author Dingenis Sieger Sinke
 * @version 1.0
 * @since 4-9-2016
 * Description...
 */
public class AsyncTransduceTask<I, O extends AbstractTransducer.Result> extends AsyncTask<Integer, Integer, O> {

    protected AbstractTransducer<I, O> transducer;

    protected I data;

    protected OnTransduceCompletedCallback<O> callback;

    public void setTransducer(AbstractTransducer<I, O> transducer) {
        this.transducer = transducer;
    }

    public void setData(I data) {
        this.data = data;
    }

    public void setCallback(OnTransduceCompletedCallback<O> callback) {
        this.callback = callback;
    }

    @Override
    protected O doInBackground(Integer... integers) {
        return transducer.generate(data);
    }

    @Override
    protected void onPostExecute(O output) {
        if(output != null && callback != null) {
            callback.onTransduceCompleted(output);
        }
    }

    public interface OnTransduceCompletedCallback<O extends AbstractTransducer.Result> {
        void onTransduceCompleted(O result);
    }
}

File: app/src/main/java/nl/rsdt/japp/jotial/maps/management/transformation/async/AsyncTransduceTask.kt
package nl.rsdt.japp.jotial.maps.management.transformation.async

import android.os.AsyncTask

import nl.rsdt.japp.jotial.maps.management.transformation.AbstractTransducer

/**
 * @author Dingenis Sieger Sinke
 * @version 1.0
 * @since 4-9-2016
 * Description...
 */
class AsyncTransduceTask<I, O : AbstractTransducer.Result> : AsyncTask<Int, Int, O>() {

    protected var transducer: AbstractTransducer<I, O>

    protected var data: I

    protected var callback: OnTransduceCompletedCallback<O>? = null

    fun setTransducer(transducer: AbstractTransducer<I, O>) {
        this.transducer = transducer
    }

    fun setData(data: I) {
        this.data = data
    }

    fun setCallback(callback: OnTransduceCompletedCallback<O>) {
        this.callback = callback
    }

    protected override fun doInBackground(vararg integers: Int): O {
        return transducer.generate(data)
    }

    override fun onPostExecute(output: O?) {
        if (output != null && callback != null) {
            callback!!.onTransduceCompleted(output)
        }
    }

    interface OnTransduceCompletedCallback<O : AbstractTransducer.Result> {
        fun onTransduceCompleted(result: O)
    }
}
